lightning-background-processor/src/lib.rs

```diff
@@ -152,11 +155,27 @@ impl BackgroundProcessor {
 					log_trace!(logger, "Terminating background processor.");
 					return Ok(());
 				}
-				if current_time.elapsed().as_secs() > FRESHNESS_TIMER {
-					log_trace!(logger, "Calling ChannelManager's and PeerManager's timer_tick_occurred");
+				if last_freshness_call.elapsed().as_secs() > FRESHNESS_TIMER {
+					log_trace!(logger, "Calling ChannelManager's timer_tick_occurred");
 					channel_manager.timer_tick_occurred();
+					last_freshness_call = Instant::now();
+				}
+				if last_ping_call.elapsed().as_secs() > PING_TIMER * 2 {
+					// On various platforms, we may be starved of CPU cycles for several reasons.
+					// eg on iOS, if we've been in the background, we will be entirely paused.
+					// Similarly, if we're on a desktop platform and the device has been asleep, we
+					// may not get any cycles.
+					// In any case, if we've been entirely paused for more than double our ping
+					// timer, we should have disconnected all sockets by now (and they're probably
+					// dead anyway), so disconnect them by calling `timer_tick_occurred()` twice.
+					log_trace!(logger, "Awoke after more than double our ping timer, disconnecting peers.");
+					peer_manager.timer_tick_occurred();
+					peer_manager.timer_tick_occurred();
```

Comment:
It reads to me as odd to disconnect peers this way, rather than explicitly through some `disconnect_peer()` or so API

Reply:
Right, to some extent I agree, but at the same times its what has happened - the timer should have fired twice, so that's just what we do. The `PeerManager` could in the future tweak how it handles timer events and I think this would be more correct.